File: Algorithms-DataStructures-in-Java/BST.java
//Program to implement Binary Search Tree

public class BST<Key extends Comparable<Key>, Value> {

	Node root;

	private class Node {
		private Value value;
		private Key key;
		private Node left, right;
		private int count;

		public Node(Key key, Value value) {
			this.value = value;
			this.key = key;
			this.count = 1;
		}

	}

	public int size(Node x) {
		if (x == null)
			return 0;
		// System.out.println(x.count);
		return x.count;
	}

	public Value get(Key k) {

		Node x = root;

		while (x != null) {

			int cmp = k.compareTo(x.key);

			if (cmp < 0)
				x = x.left;
			else if (cmp > 0)
				x = x.right;
			else
				return x.value;
		}
		return null;
	}

	public void put(Key key, Value value) {
		root = put(root, key, value);
	}

	public Node put(Node x, Key key, Value value) {

		if (x == null)
			return new Node(key, value);

		int cmp = key.compareTo(x.key);

		if (cmp < 0)
			x.left = put(x.left, key, value);
		else if (cmp > 0)
			x.right = put(x.right, key, value);
		else
			x.value = value;
		x.count = 1 + size(x.left) + size(x.right);
		return x;
	}

	public Node min(Node x) {

		while (x.left != null)
			x = x.left;
		return x;
	}

	public Node max(Node x) {

		while (x.right != null)
			x = x.right;
		return x;
	}

	public Node deleteMin(Node x) {
		if (x.left == null)
			return x.right;
		x.left = deleteMin(x.left);
		return x;
	}

	public void delete(Key key) {
		root = delete(root, key);
	}

	public Node delete(Node x, Key key) {
		if (x == null)
			return null;

		int cmp = key.compareTo(x.key);
		if (cmp < 0)
			x.left = delete(x.left, key);
		else if (cmp > 0)
			x.right = delete(x.right, key);
		else {
			if (x.right == null)
				return x.left;

			Node t = x;
			x = min(x.right);
			x.right = deleteMin(t.right);
			x.left = t.left;

		}

		return x;
	}

	public void inorder() {
		inorder(root);
	}

	public void inorder(Node x) {
		if (x == null)
			return;
		inorder(x.left);
		System.out.print(x.key + "," + x.count + "  ");
		inorder(x.right);

	}

	public Key floor(Key key) {
		Node x = floor(root, key);
		if (x == null)
			return null;
		return x.key;
	}

	public Node floor(Node x, Key key) {
		if (x == null)
			return null;
		int cmp = key.compareTo(x.key);

		if (cmp == 0)
			return x;

		if (cmp < 0)
			return floor(x.left, key);

		Node t = floor(x.right, key);
		if (t != null)
			return t;
		return x;
	}

	public int rank(Key key) {
		return rank(root, key);
	}

	public int rank(Node x, Key key) {

		if (x == null)
			return 0;
		int cmp = key.compareTo(x.key);
		if (cmp < 0)
			return rank(x.left, key);
		else if (cmp > 0)
			return 1 + size(x.left) + rank(x.right, key);
		else
			return size(x.left);

	}

	public static void main(String[] args) {

		BST<Integer, Integer> obj = new BST<>();
		obj.put(50, 30);
		obj.put(10, 10);
		obj.put(70, 20);
		obj.put(55, 50);
		obj.put(90, 40);
		obj.put(53, 60);
		obj.put(65, 60);
		obj.put(75, 60);
		// System.out.println(obj.size(obj.root));
		 System.out.println(obj.rank(70));
		// obj.delete(70);
		// obj.delete(50);
		obj.inorder();

	}

}

File: Algorithms-DataStructures-in-C/BST.c
//#include <stdio.h>
//#include <stdlib.h>
//
//struct node {
//	int key;
//	int value;
//	int count;
//	struct node *left, *right;
//};
//
//struct node *root = NULL;
//
//int size(struct node *x) {
//	if (x == NULL)
//		return 0;
//	return x->count;
//}
//
//struct node* floor12(struct node* x, int key) {
//	if (x == NULL)
//		return NULL;
//
//	if (key == x->key)
//		return x;
//
//	if (key < x->key)
//		return floor12(x->left, key);
//
//	struct node *t = floor12(x->right, key);
//	if (t != NULL)
//		return t;
//	else
//		return x;
//}
//
//int floorMain(int key) {
//	struct node* t = floor12(root, key);
//	if (t == NULL)
//		return 0;
//	return t->key;
//}
//
//struct node* insert(struct node *x, int key, int value) {
//
//	if (x == NULL) {
//		struct node *newnode = malloc(sizeof(struct node));
//		newnode->key = key;
//		newnode->value = value;
//		newnode->left = NULL;
//		newnode->right = NULL;
//		newnode->count = 1;
//		return newnode;
//	}
//
//	if (key < x->key)
//		x->left = insert(x->left, key, value);
//	else if (key > x->key)
//		x->right = insert(x->right, key, value);
//	else if (key == x->key)
//		x->value = value;
//
//	x->count = 1 + size(x->left) + size(x->right);
//	return x;
//
//}
//
//void insertMain(int key, int value) {
//	root = insert(root, key, value);
//}
//
//void inorder(struct node *x) {
//
//	if (x == NULL)
//		return;
//
//	inorder(x->left);
//	printf("%d  ", x->key);
//	inorder(x->right);
//}
//
//void inorderMain() {
//	inorder(root);
//}
//
//struct node* min(struct node *x) {
//	while (x->left != NULL)
//		x = x->left;
//	return x;
//}
//
//struct node* deleteMin(struct node *x) {
//	if (x->left == NULL)
//		return x->right;
//	x->left = deleteMin(x->left);
//	return x;
//}
//
//struct node* delete(struct node *x, int key) {
//	if (x == NULL)
//		return NULL;
//
//	if (key < x->key)
//		x->left = delete(x->left, key);
//	else if (key > x->key)
//		x->right = delete(x->right, key);
//	else if (key == x->key) {
//		if (x->right == NULL)
//			return x->left;
//
//		struct node *t = x;
//		x = min(x->right);
//		x->right = deleteMin(t->right);
//		x->left = t->left;
//
//	}
//
//	return x;
//}
//
//void deleteMain(int key) {
//	root = delete(root, key);
//}
//
//int rankMain(int key) {
//	return rank(root, key);
//}
//
//int rank(struct node* x, int key) {
//	if (x == NULL)
//		return 0;
//
//	if (key < x->key)
//		return rank(x->left, key);
//	else if (key > x->key)
//		return 1 + size(x->left) + rank(x->right, key);
//	else if (key == x->key)
//		return size(x->left);
//}
//
//int main(int argc, char **argv) {
//	insertMain(19, 100);
//	insertMain(13, 100);
//	insertMain(10, 100);
//	insertMain(20, 100);
//	insertMain(5, 100);
//	insertMain(23, 100);
//	insertMain(28, 100);
//	insertMain(16, 100);
//	insertMain(27, 100);
//	insertMain(9, 100);
//	insertMain(4, 100);
//	insertMain(22, 100);
//	insertMain(17, 100);
//	insertMain(30, 100);
//	insertMain(40, 100);
//
//	printf("%d", rankMain(21));
//
//	return 0;
//
//}
